File: code/SegmentTree.cpp
struct SegmentTree{

	int rank, capacity, length;
	int *input, *tree;

    SegmentTree() {}
	void init(int* input, int length){
		this->input = input;
		this->length = length;
		rank = 1;
		while((1<<rank++) < length);
		capacity = 1<<(rank-1);
		tree = new int[capacity << 1];
		build(1, capacity, capacity<<1);
	}

	~SegmentTree(){
        delete[] tree;
	}

	int build(int index, int left, int right){
		if(index >= left){
			return tree[index] = getInput(index);
		}
		int middle = (left+right) >> 1;
		int left_value = build(lc(index), left, middle);
		int right_value = build(rc(index), middle, right);
		return tree[index] = max(left_value, right_value);
	}

	int query(int start, int finish){
		return query(1, capacity, capacity<<1, capacity+start, capacity+finish+1);
	}

	int query(int index, int left, int right, int start, int finish){
		if(left == start && right == finish) return tree[index];
		int middle = (left+right) >> 1;
		if(finish <= middle) return query(lc(index), left, middle, start, finish);
		if(start >= middle) return query(rc(index), middle, right, start, finish);
		int left_value = query(lc(index), left, middle, start, middle);
		int right_value = query(rc(index), middle, right, middle, finish);
		return max(left_value, right_value);
	}

	int getInput(int index){
		index -= capacity;
		if(index < length) return input[index];
		return 0;
	}

	int lc(int x){
		return x<<1;
	}

	int rc(int x){
		return (x<<1)+1;
	}
};

File: code/SegmentTree.java
public class SegmentTree{

    int[] input;
    Entry[] tree;
    int rank, capacity;

    SegmentTree(int[] input){
        this.input = input;
        rank = 1;
        while(1<<(rank++) < input.length);
        capacity = ((1<<rank)>>1);
        tree = new Entry[1<<rank];
        build(1, 0, capacity);
    }

    int operate(int resultL, int resultR){
        return Math.max(resultL, resultR);
    }

    int build(int index, int left, int right){
        Entry now = tree[index] = new Entry(left, right, index);
        if(left+1 == right){
            if(left >= input.length) return now.value = 0;
            return now.value = input[left]; 
        }
        int middle = (left+right) >> 1;
        return now.value = operate(build(lc(index), left, middle), build(rc(index), middle, right));
    }

    int query(int start, int finish){
        return query(1, start, finish);
    }

    int query(int index, int start, int finish){
        //System.out.println("query "+index+", "+start+", "+finish);
        if(tree[index].lb == start && tree[index].rb == finish) return tree[index].value;
        int middle = (tree[index].lb+tree[index].rb) >> 1;
        if(finish <= middle) return query(lc(index), start, finish);
        else if(middle <= start) return query(rc(index), start, finish);
        else{
            return operate(query(lc(index), start, middle), query(rc(index), middle, finish));
        }
    }

    void update(int target, int value){
        int index = target+capacity;
        tree[index].value = value;
        maintain(index>>1);
    }

    void maintain(int index){
        tree[index].value = operate(tree[lc(index)].value, tree[rc(index)].value);
        if(index == 1) return;
        maintain(index>>1);
    }

    int lc(int x){
        return x<<1;
    }

    int rc(int x){
        return (x<<1)+1;
    }

    class Entry{

        int lb, rb, id; //Left Bound, Right Bound and index in Array
        int value;

        Entry(int lb, int rb, int id){
            this.lb = lb;
            this.rb = rb;
            this.id = id;
            value = -1;
        }

    }

}
